File: apps/web/lib/domain/class-groups.ts
import type {
  ScheduledLesson,
  SnapshotAssignment,
  TeachingGroup,
} from "./contracts";

const PARALLEL_GROUP_ORDER: Array<Exclude<TeachingGroup, "WHOLE">> = [
  "GROUP_1",
  "GROUP_2",
  "GROUP_3",
];

export function assignmentClassIds(assignment: SnapshotAssignment): string[] {
  return [
    ...new Set([
      assignment.class_id,
      ...(assignment.additional_class_ids ?? []),
    ]),
  ];
}

export function lessonClassIds(lesson: ScheduledLesson): string[] {
  return [
    ...new Set([lesson.class_id, ...(lesson.additional_class_ids ?? [])]),
  ];
}

function parallelKey(assignment: SnapshotAssignment): string {
  if (assignment.parallel_key) return assignment.parallel_key;

  const normalizedId = assignment.id.toLocaleLowerCase("cs-CZ");
  if (normalizedId.includes("-rot-") || normalizedId.includes("-rotation-")) {
    return `rotation-id:${normalizedId.replace(/-(g1|g2)$/i, "")}`;
  }
  return `subject:${assignment.subject_id}`;
}

export function parallelAssignmentGroups(
  assignments: SnapshotAssignment[],
): SnapshotAssignment[][] {
  const grouped = new Map<
    string,
    Partial<Record<Exclude<TeachingGroup, "WHOLE">, SnapshotAssignment[]>>
  >();

  for (const assignment of assignments) {
    if (assignment.group === "WHOLE") continue;
    const key = `${assignmentClassIds(assignment).sort().join("|")}::${parallelKey(assignment)}`;
    const groups = grouped.get(key) ?? {};
    const current = groups[assignment.group] ?? [];
    groups[assignment.group] = [...current, assignment];
    grouped.set(key, groups);
  }

  return [...grouped.values()].flatMap((groups) => {
    const present = PARALLEL_GROUP_ORDER.flatMap((group) => {
      const items = [...(groups[group] ?? [])].sort((a, b) =>
        a.id.localeCompare(b.id),
      );
      return items.length === 1 ? [items[0]!] : [];
    });
    const rawCount = PARALLEL_GROUP_ORDER.reduce(
      (total, group) => total + (groups[group]?.length ?? 0),
      0,
    );
    return present.length >= 2 && present.length === rawCount ? [present] : [];
  });
}

export function parallelAssignmentPairs(
  assignments: SnapshotAssignment[],
): Array<[SnapshotAssignment, SnapshotAssignment]> {
  return parallelAssignmentGroups(assignments).flatMap((group) =>
    group.length === 2 &&
    group[0]?.group === "GROUP_1" &&
    group[1]?.group === "GROUP_2"
      ? [[group[0], group[1]] as [SnapshotAssignment, SnapshotAssignment]]
      : [],
  );
}

export function rotationAssignmentLegs(
  assignments: SnapshotAssignment[],
): Array<{
  rotationKey: string;
  leg1: [SnapshotAssignment, SnapshotAssignment];
  leg2: [SnapshotAssignment, SnapshotAssignment];
}> {
  const rotations = new Map<
    string,
    Partial<Record<1 | 2, [SnapshotAssignment, SnapshotAssignment]>>
  >();

  for (const pair of parallelAssignmentPairs(assignments)) {
    const [left, right] = pair;
    if (
      !left.rotation_key ||
      left.rotation_key !== right.rotation_key ||
      left.rotation_leg == null ||
      left.rotation_leg !== right.rotation_leg
    ) {
      continue;
    }
    const leg = left.rotation_leg as 1 | 2;
    const current = rotations.get(left.rotation_key) ?? {};
    current[leg] = pair;
    rotations.set(left.rotation_key, current);
  }

  return [...rotations.entries()]
    .filter(
      (
        entry,
      ): entry is [
        string,
        {
          1: [SnapshotAssignment, SnapshotAssignment];
          2: [SnapshotAssignment, SnapshotAssignment];
        },
      ] => Boolean(entry[1][1] && entry[1][2]),
    )
    .sort(([left], [right]) => left.localeCompare(right))
    .map(([rotationKey, legs]) => ({
      rotationKey,
      leg1: legs[1],
      leg2: legs[2],
    }));
}

export function classRequiredWeeklyPeriods(
  assignments: SnapshotAssignment[],
): Map<string, number> {
  const totals = new Map<string, number>();
  const groupedIds = new Set<string>();

  for (const group of parallelAssignmentGroups(assignments)) {
    group.forEach((assignment) => groupedIds.add(assignment.id));
    const weeklyPeriods = Math.max(...group.map((item) => item.weekly_periods));
    for (const classId of assignmentClassIds(group[0]!)) {
      totals.set(classId, (totals.get(classId) ?? 0) + weeklyPeriods);
    }
  }

  for (const assignment of assignments) {
    if (groupedIds.has(assignment.id)) continue;
    for (const classId of assignmentClassIds(assignment)) {
      totals.set(
        classId,
        (totals.get(classId) ?? 0) + assignment.weekly_periods,
      );
    }
  }

  return totals;
}

File: apps/solver/app/class_groups.py
import re
from collections import defaultdict

from app.models import Assignment, ScheduledLesson, TeachingGroup

PARALLEL_GROUP_ORDER = (
    TeachingGroup.GROUP_1,
    TeachingGroup.GROUP_2,
    TeachingGroup.GROUP_3,
)


def assignment_class_ids(assignment: Assignment) -> tuple[str, ...]:
    return tuple(dict.fromkeys([assignment.class_id, *assignment.additional_class_ids]))


def lesson_class_ids(lesson: ScheduledLesson) -> tuple[str, ...]:
    return tuple(dict.fromkeys([lesson.class_id, *lesson.additional_class_ids]))


def _parallel_key(assignment: Assignment) -> str:
    if assignment.parallel_key:
        return assignment.parallel_key

    normalized_id = assignment.id.lower()
    if "-rot-" in normalized_id or "-rotation-" in normalized_id:
        return f"rotation-id:{re.sub(r'-(g1|g2)$', '', normalized_id)}"

    return f"subject:{assignment.subject_id}"


def parallel_assignment_groups(assignments: list[Assignment]) -> list[list[Assignment]]:
    grouped: dict[
        tuple[tuple[str, ...], str],
        dict[TeachingGroup, list[Assignment]],
    ] = defaultdict(lambda: defaultdict(list))

    for assignment in assignments:
        if assignment.group == TeachingGroup.WHOLE:
            continue
        key = (
            tuple(sorted(assignment_class_ids(assignment))),
            _parallel_key(assignment),
        )
        grouped[key][assignment.group].append(assignment)

    result: list[list[Assignment]] = []
    for groups in grouped.values():
        present: list[Assignment] = []
        raw_count = 0
        for group in PARALLEL_GROUP_ORDER:
            items = sorted(groups.get(group, []), key=lambda item: item.id)
            raw_count += len(items)
            if len(items) == 1:
                present.append(items[0])
        if len(present) >= 2 and len(present) == raw_count:
            result.append(present)
    return result


def parallel_assignment_pairs(
    assignments: list[Assignment],
) -> list[tuple[Assignment, Assignment]]:
    result: list[tuple[Assignment, Assignment]] = []
    for group in parallel_assignment_groups(assignments):
        if (
            len(group) == 2
            and group[0].group == TeachingGroup.GROUP_1
            and group[1].group == TeachingGroup.GROUP_2
        ):
            result.append((group[0], group[1]))
    return result


def rotation_assignment_legs(
    assignments: list[Assignment],
) -> list[tuple[str, tuple[Assignment, Assignment], tuple[Assignment, Assignment]]]:
    by_rotation: dict[str, dict[int, tuple[Assignment, Assignment]]] = defaultdict(dict)
    for left, right in parallel_assignment_pairs(assignments):
        if (
            left.rotation_key
            and left.rotation_key == right.rotation_key
            and left.rotation_leg is not None
            and left.rotation_leg == right.rotation_leg
        ):
            by_rotation[left.rotation_key][left.rotation_leg] = (left, right)

    result: list[tuple[str, tuple[Assignment, Assignment], tuple[Assignment, Assignment]]] = []
    for rotation_key, legs in sorted(by_rotation.items()):
        if 1 in legs and 2 in legs:
            result.append((rotation_key, legs[1], legs[2]))
    return result


def class_required_weekly_periods(assignments: list[Assignment]) -> dict[str, int]:
    totals: dict[str, int] = defaultdict(int)
    grouped_ids: set[str] = set()

    for group in parallel_assignment_groups(assignments):
        grouped_ids.update(item.id for item in group)
        weekly_periods = max(item.weekly_periods for item in group)
        for class_id in assignment_class_ids(group[0]):
            totals[class_id] += weekly_periods

    for assignment in assignments:
        if assignment.id in grouped_ids:
            continue
        for class_id in assignment_class_ids(assignment):
            totals[class_id] += assignment.weekly_periods

    return dict(totals)
